### ISSUE TITLE
supervisor shutdown is slow

### ISSUE THREAD

**Issue by palmkevin** [Supervisor/supervisor#326]
I know that the supervisor shutdown-time depends on the time the different processes take to exit.

But it seems that the supervisor does not send the SIGTERM signal to all its processes the moment the shutdown is requested.
Does this have a special reason? I'd prefer that all processes receive the SIGTERM event the moment the supervisor receives the shutdown command, so that slow-terminating processes can do their shutdown tasks in parallel. This would speed-up my supervisor shutdown a lot!

**Comment by kevinkreiser** [comment]
@mnaberez I think the OP was referring to something a little more complex than the test you provided. Let me elaborate:

Consider the following application `work.py` which listens for SIGTERM and upon hearing it, cleans up for 5 seconds before gracefully shutting down:

```python
#!/usr/bin/env python
import signal
import time
import threading
import sys

def cleanup(sig, frame):
  print('cleaning')
  time.sleep(5)
  global shutdown
  shutdown = True

def work(name):
  global shutdown
  while not shutdown:
    print('%s is working' % name)
    time.sleep(1)

if __name__ == '__main__':
  global shutdown
  shutdown = False
  signal.signal(signal.SIGTERM, cleanup)
  x = threading.Thread(target=work, args=(sys.argv[1],))
  x.start()
  x.join()
```

Then consider the following `supervisord.conf` in which we run multiple separate instances of the above application with different arguments (seems contrived but in real life you can imagine a bunch of different executables):

```ini
[supervisord]
logfile = /tmp/supervisord.log
loglevel = info

[inet_http_server]
port = 127.0.0.1:9001

[supervisorctl]
serverurl = http://127.0.0.1:9001

[rpcinterface:supervisor]
supervisor.rpcinterface_factory = supervisor.rpcinterface:make_main_rpcinterface

[program:foo]
command = ./work.py foo
process_name = foo

[program:bar]
command = ./work.py bar
process_name = bar
```

When you run this with `supervisord -n -c supervisord.conf` and you send SIGTERM to supervisord, you get the following output:

```
2021-03-01 19:16:27,696 INFO RPC interface 'supervisor' initialized
2021-03-01 19:16:27,696 CRIT Server 'inet_http_server' running without any HTTP authentication checking
2021-03-01 19:16:27,696 INFO supervisord started with pid 344951
2021-03-01 19:16:28,698 INFO spawned: 'bar' with pid 344953
2021-03-01 19:16:28,699 INFO spawned: 'foo' with pid 344954
2021-03-01 19:16:29,701 INFO success: bar entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2021-03-01 19:16:29,701 INFO success: foo entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2021-03-01 19:16:34,709 WARN received SIGTERM indicating exit request
2021-03-01 19:16:34,709 INFO waiting for bar, foo to die
2021-03-01 19:16:37,712 INFO waiting for bar, foo to die
2021-03-01 19:16:39,736 INFO stopped: foo (exit status 0)
2021-03-01 19:16:40,738 INFO waiting for bar to die
2021-03-01 19:16:43,741 INFO waiting for bar to die
2021-03-01 19:16:45,737 INFO stopped: bar (exit status 0)
```

The problem the OP seems to be mentioning is that supervisord sends SIGTERM to each program sequentially starting at the last one defined in the config and working its way up to the first one but it waits for the current one to exit before sending SIGTERM to the next one. We end up having to wait 10 seconds for everything to properly shutdown we should have only had to wait for 5 or 6 seconds.

Basically its doing something like this:

```python
for program in programs:
  program.terminate() # send signal
  wait_for_terminated(program, timeout) # wait for death
  # make sure its dead
  if program.alive():
    program.kill()
```

When users expect it to be doing something like this:

```python
# send all seignals
for program in programs:
  program.terminate()
# wait for all deaths
for program in programs:
  wait_for_terminated(program)
  # make sure its dead
  if program.alive():
    program.kill()
```

Can we re-open the issue, or should I open another issue to tackle the problem of multiple programs not being stopped at the same time? Or perhaps there is an existing config option in supervisord that I am just not aware of? I looked at the code to see if I could make a PR but haven't figured out exactly where supervisord is sending the signals to the child processes yet.

**Comment by mnaberez** [comment]
> Can we re-open the issue, or should I open another issue to tackle the problem of multiple programs not being stopped at the same time?

Please do not open more issues for it, it is described in #723 (still open) and probably others.

**Comment by kevinkreiser** [comment]
@mnaberez no worries, sounds good.
